Fix LLVM fuzzer sample for native_sim

aescolar opened the pull request:
Get the fuzzer sample working on native_sim[//64]

Move the LLVM fuzzing specific code out of the board main file and into the sample.
That way we avoid needing to duplicate it for native_sim and avoid having a very adhoc interface between the fuzzer test
and runner code.

Note that the plan is to deprecate native_posix in 4.0 and remove it in 4.1 the ifdef'ing in the sample for native_posix will disappear at that time.

aescolar committed:
native_simulator: Rework fuzzing support

The fuzzing support in the native_simulator had not got
to work yet. Let's remove it, and instead let the
test application handle it.

For this we make public the functions which initialize
the simulation and advance time, so the fuzzing tests
can use it.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
sample fuzzer: Move fuzzer specific code to sample and fix for native_sim

Move the LLVM fuzzing specific code out of the board main
file and into the sample.
That way we avoid needing to duplicate it for native_sim and
avoid having a very adhoc interface between the fuzzer test
and runner code.

Also ensure it works for native_sim and not just native_posix

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
samples fuzzer: Update README as it works on native_sim too

Update the readme as the sample works on native_sim,
and native_posix will be deprecated soon.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
arch posix: Correct ARCH_POSIX_LIBFUZZER kconfig help message

This option can be used now also with native_sim
and seems to work fine with both 32 and 64 bit targets.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar commented:
@andyross I was hoping you could give a review to this change as original author of the llvm fuzzing support

andyross reviewed (commented):
Note that this is going to break the SOF fuzzing layer, which they have running in CI, as it uses the same interrupt-based "outer main loop".  Add @marc-hb & @cujomalainey for visibility.  Obviously SOF is going to need to move to native_sim too (which is going to be inevitably confusing, as the platform layer on that end is named "posix", heh), but we should try to do it on that side first before we break it upstream.

Also it looks like this misses the related kconfigs (CONFIG_ARCH_POSIX_FUZZ_xxx) which got left in the old posix arch?

Maybe a better scheme if you don't want this in the arch would be to either:

* Document the `nsi_exec_for()` function as an "Official Entry Point" and add some docs somewhere on how to use it.

* Move the code here somewhere under lib/ so it can be shared with apps.

I'm neutral on whether or not this stuff can/should be cut and pasted between the sample and the downstream users.  It's not a lot of code.

aescolar commented:
Thank you very much @andyross 

> this misses the related kconfigs 

Correct. I guess we can keep them in the sample with the same name.

> Document the nsi_exec_for() function as an "Official Entry Point" and add some docs somewhere on how to use it.

I think this is probably better. (We need an interface for this, and nsi_init() & nsi_exec_for() seems a bit cleaner and maybe more generally useful for other cases)

> I'm neutral on whether or not this stuff can/should be cut and pasted between the sample and the downstream users. It's not a lot of code.

It feels like it is so little and adhoc that creating a library would be mostly overhead.

aescolar committed:
arch posix fuzzing: Move kconfig options to sample

Move the kconfig options used to configure the interrupt
and wait time to the sample which uses them instead of
having them in the architecture code.
This options are very particular for this sample and not
really an API.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
release note/migration guide: Mention fuzzing changes

The fuzzing support only worked when targetting the
native_posix board, and in a way which was very coupled
with the fuzzing sample.
This has now been refactored, so those who may have
relied on hooking to the board in the same way as this
sample will need to adapt.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

andyross reviewed (dismissed):
This looks clean, and won't break the existing client in SOF.  I'll look at getting SOF ported to native_sim as soon as I'm back in Portland next week.

kartben commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -303,5 +303,11 @@ Architectures
+* POSIX arch:
+
+  * LLVM fuzzing support has been refactored. A test application now needs to provide its own
+    LLVMFuzzerTestOneInput() hook instead of relaying on a board provided one. Check
```
```suggestion
    ``LLVMFuzzerTestOneInput()`` hook instead of relying on a board provided one. Check
```

kartben commented on arch/posix/Kconfig:
```diff
@@ -22,41 +22,19 @@ config ARCH_POSIX_RECOMMENDED_STACK_SIZE
 	bool "Build fuzz test target"
 	help
-	  Build the posix app as a LLVM libfuzzer target.  Requires
+	  Build as a LLVM libfuzzer target. Requires
```
```suggestion
	  Build as an LLVM libfuzzer target. Requires
```

kartben reviewed (dismissed):
Very minor typo that I frankly wouldn't have bothered reporting if one wasn't slightly more visible for the end user due to being in the migration guide :)

aescolar commented:
Thanks @kartben , the typos you pointed are fixed now

aescolar commented:
@andyross compared to your last review, this has just been rebased and typos fixed. You are welcome to reapprove.

marc-hb commented:
Unless there are git conflicts to solve, it's much, MUCH easier to review when the PR is NOT rebased
https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#workflow-suggestions-that-help-reviewers

Almost every CI is going to test pull/71378/merge anyway, never pull/71378/head (check the CI logs)

aescolar commented:
> Unless there are git conflicts to solve, it's much, MUCH easier to review when the PR is NOT rebased

@marc-hb the typo fixes ([7088003](https://github.com/zephyrproject-rtos/zephyr/compare/0a14a4fc2e1a5ab9443a0c407151cba0cecee97e..7088003fbb118fff36f29d89557a133360a6957f)) and rebase [2487ef8](https://github.com/zephyrproject-rtos/zephyr/compare/7088003fbb118fff36f29d89557a133360a6957f..2487ef8f8f9fc8bc504500ed8fe5a89eca44fac0) went in two separate pushes. So you can ignore the rebase all together.

andyross reviewed (approved):
Refresh +1

(I hate that policy.  Standard note to committers: feel free to assume a +1 from me given a previous approval and a promise from the submitter that there are no significant changes.  I promise I'm loud and obvious when I actually don't want something to merge.)

kv2019i commented:
FYI As discussed in the comments, SOF fuzzer is now broken https://github.com/thesofproject/sof/pull/9098 , filed https://github.com/thesofproject/sof/issues/9101 to track

andyross commented:
FWIW this is mostly on me.  I got confused about what had been agreed and didn't review the new version carefully enough for breaking changes.  Easy enough to fix downstream though, modulo the CI breakage goof.
